### PR TITLE
perf: bind deterministic embedding vector copy

### PR DESCRIPTION
**Pull request by Zigfreidish**
## Summary
- Bind `list.copy` once in `DeterministicEmbeddingRuntime.embed_inputs(...)` so duplicate cached embedding vectors avoid repeated per-iteration method lookup.
- Add a regression test proving duplicate inputs still return distinct vector lists and only unique texts invoke the backend.
- Document the slice and verification plan under `docs/plans/2026-05-22-embedding-vector-copy-binding.md`.

## Plan or Spec
- `docs/plans/2026-05-22-embedding-vector-copy-binding.md`
- Registered PR-scoped probe: `deterministic-embedding-duplicate-input-cache` in `infra/perf/pr_scoped_probes.json`.

## Commands Run
- `PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" uv run --project services/mlx-worker-python pytest -q services/mlx-worker-python/tests/test_embedding_runtime.py services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_scope_report_selects_deterministic_embedding_probe services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_registered_probes_expose_focused_commands services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_deterministic_embedding_duplicate_probe_script_emits_metrics` → `17 passed in 1.76s`
- `PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" uv run --project services/mlx-worker-python coverage run -m pytest -q services/mlx-worker-python/tests/test_embedding_runtime.py services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_scope_report_selects_deterministic_embedding_probe services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_registered_probes_expose_focused_commands services/mlx-worker-python/tests/test_pr_scoped_performance.py::test_deterministic_embedding_duplicate_probe_script_emits_metrics && PYTHONPATH="$PWD:$PWD/services/mlx-worker-python" uv run --project services/mlx-worker-python coverage json -o coverage.json && python3 scripts/changed_scope_coverage.py --coverage-json coverage.json services/mlx-worker-python/worker/runtime/deterministic_embedding_runtime.py services/mlx-worker-python/tests/test_embedding_runtime.py services/mlx-worker-python/tests/test_pr_scoped_performance.py scripts/deterministic_embedding_duplicate_probe.py` → `17 passed in 0.87s`; changed-line coverage `100%`
- Baseline registered probe on `origin/main`, 3 runs: `elapsed_ms_mean` samples `5.115239`, `5.173643`, `5.73386`
- Candidate registered probe on this branch, 3 runs: `elapsed_ms_mean` samples `5.328371`, `5.256255`, `5.212039`
- `git diff --check`

## Coverage and Metrics
- Changed-line coverage: `100%` (`13/13` measurable changed lines covered).
- Local Linux probe comparison:
  - `old_mean=5.340914ms`
  - `new_mean=5.265555ms`
  - `delta_ms=-0.075359ms`
  - `speedup=1.0143x`
- Behavior metrics remained stable: `embed_text_calls_mean=512.0`, `input_count=8192.0`, `unique_input_count=512.0`, `checksum=2169.411765`.

## Known Gaps
- This is a small Python micro-optimization; local Linux probe improvement is modest and may be noise-sensitive.
- CI must run the registered PR-scoped performance workflow before merge.

## Evidence Checklist
- [x] Affected path has a registered PR-scoped performance probe.
- [x] Focused regression tests passed locally.
- [x] Changed-scope coverage is at least 95%.
- [x] Registered probe command ran locally on Linux for old and new revisions.
- [ ] GitHub Actions completed successfully, including PR-scoped performance validation.

### CODE REVIEW

**Comment by Zigfreidish** [comment]
Closing this slice as rejected: the registered PR-scoped performance report completed, but `deterministic-embedding-duplicate-input-cache` flagged a regression.

Evidence from the CI report:
- `elapsed_ms_mean`: base `4.522ms`, head `4.760ms`, delta `+0.238ms` / `+5.27%` → `regression`
- `embed_text_calls_mean`: unchanged at `512.000`
- Targeted tests: pass
- Coverage: pass (`100.0%`)

Local Linux probes were noise-sensitive and showed only a small apparent improvement (`old_mean=5.340914ms`, `new_mean=5.265555ms`, `speedup=1.0143x`), so the CI registered probe is the deciding evidence. This optimization will not be merged.

**Comment by Zigfreidish** [comment]
Rejected due to registered PR-scoped probe regression; not merging this performance slice.

File: services/mlx-worker-python/tests/test_embedding_runtime.py [hunk 1]
```diff
@@ -63,6 +63,20 @@ def embed_text(
         return backend.embed_text(text, dimensions)
 
 
+class CopyTrackingVector(list[float]):
+    copy_calls = 0
```
Comment (Zigfreidish) [inline services/mlx-worker-python/tests/test_embedding_runtime.py]: Class-level mutable state here (`copy_calls = 0`) is shared across all test runs. The test that uses this correctly resets it with `CopyTrackingVector.copy_calls = 0` before asserting, which works. Just worth being aware that if a future test exercises `CopyTrackingVector` without resetting (or if the suite ever runs tests in threads), counts will bleed. A lightweight alternative is to use an instance variable and track via the specific backend instance, or to use a pytest fixture for reset. Not a blocker given the current explicit reset, but something to keep in mind if the test helper grows.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: docs/plans/2026-05-22-embedding-vector-copy-binding.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+This slice is limited to `DeterministicEmbeddingRuntime.embed_inputs(...)` in the Python worker embedding runtime. The hot path handles batches with duplicate input text by reusing cached vectors and returning a defensive copy for duplicate outputs.
+
+## Registered probe
+
+The affected path is covered by the registered PR-scoped probe `deterministic-embedding-duplicate-input-cache` in `infra/perf/pr_scoped_probes.json`. That registry entry includes focused `test_command`, `coverage_command`, and `probe_command` fields and runs on Linux via `ubuntu-latest`.
+
+## Outcome
+
+Keep the duplicate-input cache on the vector object's `copy()` method. The attempted
```
Comment (Zigfreidish) [inline docs/plans/2026-05-22-embedding-vector-copy-binding.md:13]: Minor nit: the PR title ("perf: bind deterministic embedding vector copy") implies the binding landed, but this Outcome section correctly describes the reversal. Consider aligning the PR title — e.g. "test: add regression coverage for embedding vector copy isolation" — to avoid the impression that a new optimization was shipped when the net production-code diff is zero.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by Zigfreidish — COMMENTED** [review]
Overall this is clean and correct. The two new tests are well-structured: `test_embed_duplicate_input_cache_returns_distinct_vector_lists` verifies backend deduplication and mutation isolation, and `test_embed_duplicate_input_cache_uses_vector_copy_method` pins the dispatch path (vector's own `.copy()`, not `list.copy(vector)`) to prevent the regression from creeping back. Coverage is 100% on the touched lines. Two minor notes left inline — neither is a blocker. Everything looks good to merge.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by Zigfreidish** [comment]
Closing this rejected performance slice. The latest registered PR-scoped report is acceptable overall, but the slice did not provide a clear improvement: duplicate-input elapsed time was neutral/slightly slower (+1.36%) while projection allocation was only neutral/slightly faster (-0.38%). Per the one-slice performance rule, no-improvement slices should not be merged.

File: services/mlx-worker-python/worker/runtime/deterministic_embedding_runtime.py [hunk 3]
```diff
@@ -6,6 +6,28 @@
 )
 
 
+def _repeated_input_cycle_length(inputs: list[str]) -> int:
+    input_count = len(inputs)
+    if input_count < 1024:
```
Comment (Zigfreidish) [inline services/mlx-worker-python/worker/runtime/deterministic_embedding_runtime.py:11]: The `1024` minimum is a magic number. A module-level constant with a short rationale comment would make it easier to tune and explain why small batches skip the cycle scan:

```python
# Cycle detection overhead isn't worth it below this batch size.
_CYCLE_DETECTION_MIN_INPUTS = 1024
```

Then `if input_count < _CYCLE_DETECTION_MIN_INPUTS:`. Minor readability point.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: services/mlx-worker-python/worker/runtime/deterministic_embedding_runtime.py [hunk 4]
```diff
@@ -36,11 +58,23 @@ def embed_inputs(self, loaded_model, inputs: list[str]) -> list[list[float]]:
...
-        vector_cache: dict[str, list[float]] = {}
         vectors: list[list[float]] = []
-        cache_get = vector_cache.get
         append_vector = vectors.append
         embed_text = family.embed_text
+        cycle_length = _repeated_input_cycle_length(inputs)
+        if cycle_length:
+            for text in inputs[:cycle_length]:
+                vector = embed_text(backend, text, dimensions)
+                append_vector(vector)
```
Comment (Zigfreidish) [inline services/mlx-worker-python/worker/runtime/deterministic_embedding_runtime.py:68]: Intentional asymmetry worth a comment: vectors at positions `0..cycle_length-1` (the first cycle) are returned as the raw `embed_text` output without copying, while positions in repeat cycles get `vector.copy()` (line 72). This is consistent with the non-cycle path where the first occurrence of each unique text is also returned without a copy (line 83) and only cached-hit duplicates are copied (line 85).

The asymmetry is safe here because `cycle_vectors` holds the only other references and drops out of scope on `return vectors`, so the caller owns the first-cycle objects outright. A brief comment would prevent a future reader from "fixing" this into an unnecessary extra allocation:

```python
# First cycle vectors are owned outright by the caller; no copy needed.
# Repeat-cycle positions use .copy() because cycle_vectors still references
# the same objects until this function returns.
append_vector(vector)
```

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by Zigfreidish — COMMENTED** [review]
Cycle-detection logic in `_repeated_input_cycle_length` is correct: it finds the first repeated element, treats that index as the candidate cycle length, validates divisibility, then walks every subsequent chunk for an exact match. The probe assertions (`assert_cycle_detection_contract`) cover the key boundary cases. The fix to use `vector.copy()` (the instance method) instead of `list.copy(vector)` (unbound method call that bypasses subclass overrides) is the right call.

Two inline notes:
1. The `1024` threshold is a magic number — easy to extract into a named constant.
2. The first-cycle / repeat-cycle copy asymmetry is correct but non-obvious; a comment would prevent a future "helpfulness refactor" from adding an unnecessary copy.

`integration-tests` was still `in_progress` when this review was written — worth confirming it completes green before merging. Everything else looks good.

---
_Generated by [Claude Code](https://claude.ai/code)_
